### ISSUE TITLE
undefined method 

### ISSUE THREAD

**Issue by Paladin** [chef-brigade/postgresql-cookbook#376]
Not sure, this *might* be related to #361, but I got:

 ```
     NoMethodError
       -------------
       undefined method `[]' for nil:NilClass
       
       Cookbook Trace:
       ---------------
         /tmp/kitchen/cache/cookbooks/postgresql/recipes/yum_pgdg_postgresql.rb:26:in `from_file'
         /tmp/kitchen/cache/cookbooks/postgresql/recipes/client.rb:31:in `from_file'
         /tmp/kitchen/cache/cookbooks/postgresql/recipes/server.rb:22:in `from_file'
         /tmp/kitchen/cache/cookbooks/pk_components/recipes/postgres.rb:13:in `from_file'
         /tmp/kitchen/cache/cookbooks/pk_components/recipes/default.rb:11:in `from_file'
       
       Relevant File Content:
       ----------------------
       /tmp/kitchen/cache/cookbooks/postgresql/recipes/yum_pgdg_postgresql.rb:
       
        19:  # Install the "PostgreSQL RPM Building Project - Yum Repository"
        20:  
        21:  rpm_platform = node['platform']
        22:  rpm_platform_version = node['platform_version'].to_f.to_i.to_s
        23:  arch = node['kernel']['machine']
        24:  
        25:  # Download the PGDG repository RPM as a local file
        26>> remote_file "#{Chef::Config[:file_cache_path]}/#{node['postgresql']['pgdg']['repo_rpm_url'][node['postgresql']['version']][rpm_platform][rpm_platform_version][arch]['package']}" do
        27:    source "#{node['postgresql']['pgdg']['repo_rpm_url'][node['postgresql']['version']][rpm_platform][rpm_platform_version][arch]['url']}#{node['postgresql']['pgdg']['repo_rpm_url'][node['postgresql']['version']][rpm_platform][rpm_platform_version][arch]['package']}"
        28:    mode '0644'
        29:  end
        30:  
        31:  # Install the PGDG repository RPM from the local file
        32:  package (node['postgresql']['pgdg']['repo_rpm_url'][node['postgresql']['version']][rpm_platform][rpm_platform_version][arch]['package']).to_s do
        33:    provider Chef::Provider::Package::Rpm
        34:    source "#{Chef::Config[:file_cache_path]}/#{node['postgresql']['pgdg']['repo_rpm_url'][node['postgresql']['version']][rpm_platform][rpm_platform_version][arch]['package']}"
        35:    action :install
       
       Platform:
       ---------
       x86_64-linux

```
Run from ChefDK 0.19.6
using: (without these lines it builds and Ubuntu 16.04 builds even with these lines)
```
node.default['postgresql']['enable_pgdg_apt'] = true
node.default['postgresql']['enable_pgdg_yum'] = true

```
in the wrapper just before including postgresql::server

Target VM is driven by Vagrant, CentOS7.2

If I set node['postgresql']['version'] to '9.6' it works for CentOS, (but not with ubuntu) so it looks like there may be an issue with the default value for that on the redhat side?

**Comment by tas50** [comment]
@Paladin What version of this cookbook are you using?

**Comment by Paladin** [comment]
Sorry about the delay.
According to the list in berksfile.lock -- 5.1.0

Interestingly enough, when I drop the requested version back to 9.5 Ubuntu works:

```
node.default['postgresql']['version'] = '9.5'
node.default['postgresql']['password']['postgres'] = '###########'
node.default['postgresql']['enable_pgdg_apt'] = true
node.default['postgresql']['enable_pgdg_yum'] = true
node.default['postgresql']['config_pgtune']['db_type'] = 'web'

include_recipe 'postgresql::server'

```
Same lines with v9.6 fail (I'm doing some digging now to try and better characterize the CentOS error that started this, whether it fails at *any* given version, because I know it fails w/o a version.)

**Comment by Paladin** [comment]
Interesting. Version 9.2 is the CentOS 7 default version, and that version is the last to have issues. Recipe builds fine on CentOS 7 w/ versions 9.3 - 9.6.

Unfortunately the Ubuntu bug that surface seems a bit deeper. It seems to be installing 9.5 no matter what version is specified, but then tries to start whatever version was specified, so initdb fails.

**Comment by Paladin** [comment]
```
       Recipe: postgresql::client
         * apt_package[postgresql-client-9.5] action install
           - install version 9.5.5-1.pgdg16.04+1 of package postgresql-client-9.5
         * apt_package[libpq-dev] action install
           - install version 9.6.1-1.pgdg16.04+1 of package libpq-dev
       Recipe: postgresql::server_debian
         * apt_package[postgresql-9.5] action install
           - install version 9.5.5-1.pgdg16.04+1 of package postgresql-9.5

```   
This is what the log shows, when the specified version is 9.6

I think the root of the issue may lie in the different way the yum and apt packages are named, and cascade down from there. When a version is specified for CentOS, the server packages don't have the version number in the name, so they don't need to change. But when a non-default version is specified for Ubuntu/Debian, there are a lot of extra hoops to jump through, as the version number shows up everywhere. The server->packages list, for example, seems to stay at postgresql-9.5 so it installs 9.5 but issues other instructions as if 9.6 were installed.

In addition, if the server->packages attribute is overriden, then the right server package gets installed, but the template from server_conf goes pear-shaped because the directory it wants to write to -- postgresql/9.5/main -- no longer exists (it's 9.6, now) thanks to the 'dir' attribute).

Bottom line is because of these:

```
    default['postgresql']['version'] = '9.5'
    default['postgresql']['dir'] = '/etc/postgresql/9.5/main'
    default['postgresql']['server']['service_name'] = 'postgresql'
    default['postgresql']['client']['packages'] = ['postgresql-client-9.5', 'libpq-dev']
    default['postgresql']['server']['packages'] = ['postgresql-9.5']
    default['postgresql']['contrib']['packages'] = ['postgresql-contrib-9.5']

```

when you want to change the default version installed under ubuntu/debian you have to change as many as 5 attributes, not just the version number. And off the top of my head I don't see an elegant way of avoiding that ugliness without making other assumptions that might break just as easily.

Ah, well.

**Comment by Paladin** [comment]
Thread tl;dr:

When using the pgdg repos, the default version doesn't build on CentOS 7, but any version after 9.2 (the default version) does. According to [this commit](https://github.com/sous-chefs/postgresql-cookbook/commit/c5fc41315aaa8c0573683210609501cbad0237b4) there's no value for version 9.2 on CentOS 7. The default version attribute for CentOS 7 is set at 9.2, ergo this won't build properly by default in those conditions. One possible fix is upping the default to 9.3 when CentOS 7 is in play. Another is some guard code to check ensure 9.3 or better is selected in the version when CentOS 7+ and pgdg options are true. Which approach is more appealing?

The Ubuntu portion of the thread is kinda covered in the docs, so maybe isn't needed. I missed it the first 2-3 times I looked (yes I RTFM'ed, just didn't connect the part of the M where it was with the problem I was RTF-ing to solve) though, but that might just be me.

**Comment by erenfro** [comment]
Honestly, my personal thought on this matter, if pgdg is enabled, it itself should /default/ to the latest available pgdg version as defined. If a person wanted a specific version beyond that then they could specify it directly as normal. 

I hit this problem as well, and because pgdg defaulted to 9.2, which isn't defined in  the pgdg version hash attributes file specifically for centos 7, it hit the very same bug as this.

Ideally the default being the latest would keep rolling for the future rather than lingering in the past. I mean, this is PostgreSQL here, it's always been generally fully backwards compatible in itself. But like I said, if someone using this cookbook wanted pgdg, but did not want the latest, it's easy to set in the environment, role, or node attributes the specific version they'd want instead.

**Comment by Paladin** [comment]
I can see locking even default versions of this cookbook to specific pgsql versions; that way my cookbook (if I pinned my recipe versions like I'm supposed to) will consistently build the same server, with all the same versions of the software, every time it runs. Otherwise compatibility issues can arise.

Example: Given a cookbook centered around the nebulous "latest" version of pgsql. Build a pg server in the time before a big feature, say JSON, is added. Then use the same cookbook to build another server after. Now you have two machines, each built with the exact same version of your cookbook and recipes, plus the same version of the supermarket's postgresql cookbook, possibly even built within months or weeks of each other, which cannot support the same applications.

So I'd think a version number would *always* have to be specified by the recipe. But the recipe version should be bumped when new versions of pgsql come along so the default version built into it will still work. To achieve your idea, simply release a new version of the cookbook when a new version of postgresql ships. (Yes, I know I stretched the definition of 'simply' there. Forgive me?)

Doing this fixes the example above. If I create the same issue by updating my cookbook to use a later version of the postgresql cookbook, with a higher version number of the cookbook defaulting to a different version of pgsql, I would remedy the situation easily by simply running my updated cookbook against the older server to bring it up to date with the newly created server. But in the case of the 'latest' default, this wouldn't necessarily work. There would be insufficient information available for chef to decide how, or even whether it needs to update postgresql so running it against the out-of-date server will not be correct.
